artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java

```diff
@@ -81,44 +99,144 @@ public void generate(
           int validatorCount,
       @Option(
               names = {"--output-file", "-o"},
+              paramLabel = "<FILE>",
               description =
-                  "File to write validator keys to. Keys are printed to std out if not specified")
-          String outputFile) {
+                  "File to write unencrypted validator keys to. Keys are printed to std out if not specified")
+          String outputFile,
+      @Option(
+              names = {"--encrypt-keys", "-e"},
+              defaultValue = "false",
+              paramLabel = "<TRUE|FALSE>",
+              description = "Encrypt validator keys. (Default: false)",
+              arity = "1")
+          boolean encryptKeys,
+      @Option(
+              names = {"--keystore-output-dir", "-k"},
+              paramLabel = "<PATH>",
+              description =
+                  "Encrypted keystore output directory. Use current directory if not specified")
+          String keystorePath) {
+
     try (params) {
+      // generate BLS validator key and withdrawal key
+      final List<Pair<BLSKeyPair, BLSKeyPair>> validatorKeyPairs =
+          IntStream.range(0, validatorCount)
+              .mapToObj(value -> Pair.of(BLSKeyPair.random(), BLSKeyPair.random()))
+              .collect(Collectors.toList());
+
       final DepositTransactionSender sender = params.createTransactionSender();
-      final List<SafeFuture<TransactionReceipt>> futures = new ArrayList<>();
 
-      final PrintStream keyWriter;
-      if (outputFile == null || outputFile.isBlank()) {
-        keyWriter = System.out;
-      } else {
-        keyWriter = new PrintStream(new FileOutputStream(outputFile), true, StandardCharsets.UTF_8);
-      }
+      final StringBuilder yamlStringBuilder = new StringBuilder();
+      final boolean printOnStandardOut = outputFile == null || outputFile.isBlank();
+      final Path keyStoreOutputDirectory =
+          keystorePath == null || keystorePath.isBlank() ? Path.of(".") : Path.of(keystorePath);
+      final AtomicInteger counterSuffix = new AtomicInteger(0);
+      final List<SafeFuture<TransactionReceipt>> futures;
+      try {
+        futures =
+            validatorKeyPairs.stream()
+                .map(
+                    keyPair -> {
+                      final String yamlFormattedString = getYamlFormattedString(keyPair);
+                      if (encryptKeys) {
+                        generateKeystores(
+                            keyPair, keyStoreOutputDirectory, counterSuffix.incrementAndGet());
+                      } else if (printOnStandardOut) {
+                        System.out.println(yamlFormattedString);
+                      } else {
+                        yamlStringBuilder
+                            .append(yamlFormattedString)
+                            .append(System.lineSeparator());
+                      }
```

Comment:
Two issues here:
1. It's vital that we write the keys to output *before* sending the deposit transaction.  Otherwise if we fail somewhere we've just burned a lot of ETH and lost the keys forever.
2. Since the logic for writing keys has become more complex we should create a `KeyWriter` interface with 2 separate implementations - one that writes YAML to a `PrintStream` (used for all unencrypted output) and one that writes to a keystore for encrypted output.

Reply:
Yes I hear you, thats why I wrote file in finally block that should have written keys up to sendTransaction fails. Using KeyWrite interface is better idea though.

Reply:
A finally block does not guarantee it happens first.  If the process or computer crashed the key would never be written and the funds would be lost.

Reply:
We now have KeyWriter which is implemented by YamlKeysWriter and EncryptedKeyStoreWriter.

Reply:
The approach to write keys has been changed so that they get persisted first before sendTransaction is called.